### DISCUSSION TITLE
A function's multiple return values cannot be arguments to another function

### DISCUSSION THREAD

**Discussion by eptx** [vlang/v#16476] [Ideas]
Is this by design or WIP?

```
fn get2() (int, string) {
    return 0, 'a'
}

fn process2(n int, s string) string {
    return '${n}:${s}'
}

fn process3(n int, s string, s2 string) string {
    return '${n}:${s}:${s2}'
}

println(process2(1,'b'))
println(process2(get2()))
println(process3(get2(),'a'))
```
Results in ...
```

code.v:15:9: error: expected 2 arguments, but got 1
   13 | 
   14 | println(process2(1,'b'))
   15 | println(process2(get2()))
      |         ~~~~~~~~~~~~~~~~
   16 | println(process3(get2(),'a'))
code.v:16:9: error: expected 3 arguments, but got 2
   14 | println(process2(1,'b'))
   15 | println(process2(get2()))
   16 | println(process3(get2(),'a'))
      |         ~~~~~~~~~~~~~~~~~~~~
Exited with error status 1
```

**Comment by JalonSolov** [comment]
Looks like a bug, to me.  V should be smart enough to figure out how many things are being returned from what's called.  Looks like it is simply counting the number of things between the `()` instead of seeing that a function is being called.

**Comment by Wajinn** [comment]
Is it really a bug or is it a new feature request?  
It doesn't seem like it was something that was working, that doesn't work now, but rather something that had not been noticed.

Wondering if we are in the situation again where users might be confusing tuples with multiple returns? 
However, if it can work where the number of returns can count as the number of arguments, then that seems great.  Probably, most people were doing it as below.  And since many languages do not do multiple returns or tuples, it may not have been considered an option for various others, so why it wasn't talked about before.  

```
fn get2() (int, string) {
    return 0, 'a'
}

fn process2(n int, s string) string {
    return '${n} : ${s}'
}

fn process3(n int, s string, s2 string) string {
    return '${n} : ${s} : ${s2}'
}

x, y := get2()

println(process2(1,'b'))
println(process2(x, y))
println(process3(x, y, 'a'))
```

Edit: possible alternative using sum types.

```
type Values = int | string

fn get2() []Values {
    return [Values(0), Values('a')]
}

fn process2(n Values, s Values) string {
    return '${n} : ${s}'
}

fn process3(ns []Values, s2 Values) string {
    return '${ns} : ${s2}'
}

println(get2())
println(process2(get2()[0], get2()[1]))
println(process3(get2(), 'a'))
```

**Reply by eptx** [reply]
Yes. A couple weeks ago when 1st starting with V, the syntax of multiple values, was indeed confusing/inconsistent as I briefly thought it was using tuples...See below. Then wondered why no tuples as in Rust and Haskell. Now I see that V's direction is to limit itself from the functional paradigm to be more of a Go+ positioning. But does seem overly quirky and limiting that one function can return values that another function can't consume without creating interim variables...making multi-return vars its own anti-pattern in favor of returning structs instead.
```
fn dosomething(a int, b string) (int, string) was tuple-like and consistent.
```
but then the following was not...
```
return a, b // instead of return (a,b)
x, y := dosomething(...) // instead of (x, y) := dosomething(...)
```

**Reply by Wajinn** [reply]
Multiple returns are not inconsistent, just different.  Because a pet language does things a certain way (use tuple), doesn't mean all other languages should.  Each change has to be thought about carefully, else it turns into more of a negative than a positive.

The first point is that a tuple is a single item that contains multiple values, where the point of multiple returns is separate values which can/are to be treated independently and differently.  In many cases, where a tuple is used, an array or map can be used instead.  Furthermore, V can have sum type maps, which contain different types.  So, it could be argued there are good workarounds for tuples, in languages that don't have them.

As we got into with tuples and multiple returns before on discussions, in an unrelated situation, I had read up a bit on the different philosophies behind each choice.  If I remember correctly, I think that tuples were rejected by Go, because their developers thought it was already late in the development process, would cause inconsistency in the language, and would have a negative domino effect:

1. You could start to get multiple functions as arguments, which debatably could cause confusion. `process(get(), hit(), tip())`  

2. Then this could lead to tampering with multiple returns in general 
 `a, b, c, d := get(), hit(), tip()` // unclear what is matching to what

3. It was thought that multiple return to argument matching could lead to unseen and hard to trace errors.  The code for achieving the match could fail, so at times no errors for functions with the wrong number of arguments will show.

Now with that typed, it doesn't mean multiple returns to argument matching code can't work.  It is for developers to debate if it can be pulled off, and maybe some language(s) has done it (and without going tuple), but its hard to check and don't know of any off the top of my head.

**Reply by eptx** [reply]
Thanks @wajinn for the background. "Pet language", lol. My comments were in agreement you that yes, I briefly had confused V's multiple returns with tuples. The inconsistency I mention, is not that V doesn't have tuples, it was in the multiples syntax less familiar to me (parens in function definition, but not elsewhere). Again, V models after go which I'm less familiar with. Scala, Swift, Haskel, Rust use tuples for multiple return values and is just more familiar to me.

It does seem more like a bug however that V currently isn't able to pass a multi-return function as input to another function.

**Reply by Wajinn** [reply]
I kind of support your suggestion of multiple return values to arguments, in the context of it might be easier for newcomers and casuals, though maybe more as an experiment to see how it goes.

As noted by Go developers, it can lead to a negative domino effect of expectation to expand it further.  Thus, I think it arguably should be contained to only function arguments, and if the multiple return values to arguments feature proves it works reliably.  Maybe even further, the documentation may have to make it clear that multiple returns are not tuples, and people can use arrays, maps, or sum type arrays/maps too.

It also appears that Lisp, or some flavor of it, has kind of done this in their language (multiple value bind).  So its not a far out idea, and looks doable, though the developers would be the ones to weigh the benefits of it.

**Reply by dumblob** [reply]
@Wajinn did you or @eptx create an issue in the tracker already (I could not find it)? If so, please post a link here, thanks!

**Reply by eptx** [reply]
https://github.com/vlang/v/issues/16502
